src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,985 @@
+    }
+
+    return nullptr;
+}
+
+// Shuts down the hub listener (if enabled and running) for this node.
+void Esp32WiFiManager::stop_hub()
+{
+    if(hub_)
+    {
+        mdns_unpublish(hubServiceName_.c_str());
+        LOG(INFO, "[HUB] Shutting down TCP/IP listener");
+        //hub_->shutdown();
+        hub_.reset(nullptr);
+    }
+}
+
+// Creates a hub listener for this node after loading configuration details.
+void Esp32WiFiManager::start_hub()
+{
+    hubServiceName_ = cfg_.hub().service_name().read(configFd_);
+    uint16_t hub_port = CDI_READ_TRIMMED(cfg_.hub().port, configFd_);
+
+    LOG(INFO, "[HUB] Starting TCP/IP listener on port %d", hub_port);
+    hub_.reset(new GcTcpHub(stack_->can_hub(), hub_port));
+
+    // wait for the hub to complete it's startup tasks
+    while(!hub_->is_started())
+    {
+        vTaskDelay(HUB_STARTUP_DELAY);
+    }
+    mdns_publish(NULL, hubServiceName_.c_str(), hub_port);
+}
+
+// Disconnects and shuts down the uplink connector socket if running.
+void Esp32WiFiManager::stop_uplink()
+{
+    if(uplink_)
+    {
+        LOG(INFO, "[UPLINK] Disconnecting from uplink.");
+        uplink_->shutdown();
+        uplink_.reset(nullptr);
+    }
+}
+
+// Creates an uplink connector socket that will automatically add the uplink to
+// the node's hub.
+void Esp32WiFiManager::start_uplink()
+{
+    unique_ptr<SocketClientParams> params(new Esp32SocketParams(configFd_,
+        cfg_.uplink()));
+    uplink_.reset(new SocketClient(stack_->service(), stack_->executor(),
+        stack_->executor(), std::move(params),
+        std::bind(&Esp32WiFiManager::on_uplink_created, this,
+            std::placeholders::_1, std::placeholders::_2)
+    ));
+}
+
+// Converts the passed fd into a GridConnect port and adds it to the stack.
+void Esp32WiFiManager::on_uplink_created(int fd, Notifiable *on_exit)
+{
+    const bool use_select =
+        (config_gridconnect_tcp_use_select() == CONSTANT_TRUE);
+    create_gc_port_for_can_hub(stack_->can_hub(), fd, on_exit, use_select);
+}
+
+} // namespace openmrn_arduino
+
+/// Maximum number of milliseconds to wait for mDNS query responses.
+static constexpr uint32_t MDNS_QUERY_TIMEOUT = 2000;
+
+/// Maximum number of results to capture for mDNS query requests.
+static constexpr size_t MDNS_MAX_RESULTS = 10;
+
+// macro for splitting a service name since the ESPmDNS library requires the
+// service name and service protocol to be split.
+#define SPLIT_MDNS_SERVICE_NAME(service_name, service_protocol) \
+    if(service_name.length() && service_name.find('.', 0) != string::npos) \
+    { \
+        string::size_type split_loc = service_name.find('.', 0); \
+        service_protocol = service_name.substr(split_loc + 1); \
+        service_name.resize(split_loc); \
+    }
+
+// Advertises an mDNS service name for this node.
+void mdns_publish(const char *name, const char *service, uint16_t port)
+{
+    string service_name = service, protocol_name;
```

Comment:
we don't usually declare more than one variable on a single line.

Reply:
will split.

Reply:
will fix